internal/services/v1/schema.go

```diff
@@ -105,72 +118,104 @@ func (ss *schemaServer) WriteSchema(ctx context.Context, in *v1.WriteSchemaReque
+		existingObjectDefs, err := rwt.ListNamespaces(ctx)
 		if err != nil {
 			return err
 		}
 
-		existingDefMap := make(map[string]*core.NamespaceDefinition, len(existingDefs))
-		existing := strset.NewWithSize(len(existingDefs))
-		for _, existingDef := range existingDefs {
-			existingDefMap[existingDef.Name] = existingDef
-			existing.Add(existingDef.Name)
+		existingObjectDefMap := make(map[string]*core.NamespaceDefinition, len(existingObjectDefs))
+		existingObjectDefNames := util.NewSet[string]()
+		for _, existingDef := range existingObjectDefs {
+			existingObjectDefMap[existingDef.Name] = existingDef
+			existingObjectDefNames.Add(existingDef.Name)
 		}
 
 		// For each definition, perform a diff and ensure the changes will not result in any
 		// relationships left without associated schema.
-		for _, nsdef := range nsdefs {
-			if err := shared.SanityCheckExistingRelationships(ctx, rwt, nsdef, existingDefMap); err != nil {
+		for _, nsdef := range compiled.ObjectDefinitions {
+			if err := shared.SanityCheckExistingRelationships(ctx, rwt, nsdef, existingObjectDefMap); err != nil {
 				return err
 			}
 		}
-		log.Ctx(ctx).Trace().Interface("namespaceDefinitions", nsdefs).Msg("validated namespace definitions")
+		log.Ctx(ctx).Trace().Int("objectDefinitions", len(compiled.ObjectDefinitions)).Int("caveatDefinitions", len(compiled.CaveatDefinitions)).Msg("validated namespace definitions")
 
 		// Ensure that deleting namespaces will not result in any relationships left without associated
 		// schema.
-		removed := strset.Difference(existing, newDefs)
-		var checkRelErr error
-		removed.Each(func(nsdefName string) bool {
-			checkRelErr = shared.EnsureNoRelationshipsExist(ctx, rwt, nsdefName)
-			return checkRelErr == nil
-		})
-		if checkRelErr != nil {
-			return checkRelErr
+		removedObjectDefNames := existingObjectDefNames.Subtract(newObjectDefNames)
+		if err := removedObjectDefNames.ForEach(func(nsdefName string) error {
+			return shared.EnsureNoRelationshipsExist(ctx, rwt, nsdefName)
+		}); err != nil {
+			return err
+		}
+
+		// Write the new caveats.
+		if len(compiled.CaveatDefinitions) > 0 {
+			if err := rwt.WriteCaveats(compiled.CaveatDefinitions); err != nil {
+				return err
+			}
 		}
 
 		// Write the new namespaces.
-		if err := rwt.WriteNamespaces(nsdefs...); err != nil {
+		if err := rwt.WriteNamespaces(compiled.ObjectDefinitions...); err != nil {
 			return err
 		}
 
 		usagemetrics.SetInContext(ctx, &dispatchv1.ResponseMeta{
-			DispatchCount: uint32(len(nsdefs)),
+			DispatchCount: uint32(len(compiled.ObjectDefinitions) + len(compiled.CaveatDefinitions)),
 		})
 
 		// Delete the removed namespaces.
-		var removeErr error
-		removed.Each(func(nsdefName string) bool {
-			removeErr = rwt.DeleteNamespace(nsdefName)
-			return removeErr == nil
-		})
-		if removeErr != nil {
-			return removeErr
+		if err := removedObjectDefNames.ForEach(rwt.DeleteNamespace); err != nil {
```

Comment:
This transaction already does a lot, so if we can avoid roundtrips by merging the individual delete statements into a single large one, I think we should

Reply:
`DeleteNamespace` only takes in a single namespace

Reply:
Yep: https://github.com/authzed/spicedb/issues/935